Dockerfile

```diff
@@ -5,58 +5,82 @@
 
 # We don't use openssl (3.0) for now. We only install it so that users can use it.
 # Once cryptography is bumped to the latest version, openssl1.1-compat should be removed and openssl1.1-compat-dev
 # should be replaced by openssl-dev.
-RUN apk add --no-cache bash openssh ca-certificates jq curl openssl openssl1.1-compat perl git zip \
- && apk add --no-cache --virtual .build-deps gcc make openssl1.1-compat-dev libffi-dev musl-dev linux-headers \
- && apk add --no-cache libintl icu-libs libc6-compat \
- && apk add --no-cache bash-completion \
+
+RUN apk add --no-cache \
+    bash openssh ca-certificates jq curl openssl openssl1.1-compat perl git zip \
+    libintl icu-libs libc6-compat \
+    bash-completion \
  && update-ca-certificates
 
+#---------------------------------------------------------------------------------------------
+
+FROM common AS tools
+
 ARG JP_VERSION="0.1.3"
 
 RUN curl -L https://github.com/jmespath/jp/releases/download/${JP_VERSION}/jp-linux-amd64 -o /usr/local/bin/jp \
  && chmod +x /usr/local/bin/jp
 
+#---------------------------------------------------------------------------------------------
+
+FROM common AS builder
+
+# bash gcc make openssl-dev libffi-dev musl-dev - dependencies required for CLI
+
+RUN apk add --no-cache --virtual .build-deps gcc make openssl1.1-compat-dev libffi-dev musl-dev linux-headers
+
 WORKDIR azure-cli
 COPY . /azure-cli
 
 # 1. Build packages and store in tmp dir
 # 2. Install the cli and the other command modules that weren't included
-RUN ./scripts/install_full.sh \
- && cat /azure-cli/az.completion > ~/.bashrc \
- && runDeps="$( \
+RUN ./scripts/install_full.sh
+
+# Remove CLI source code from the final image and normalize line endings.
+RUN dos2unix az.completion /usr/local/bin/az
+
+#---------------------------------------------------------------------------------------------
+
+FROM common
+
+ARG CLI_VERSION
+
+# Metadata as defined at http://label-schema.org
+ARG BUILD_DATE
+
+LABEL maintainer="Microsoft" \
+      org.label-schema.schema-version="1.0" \
+      org.label-schema.vendor="Microsoft" \
+      org.label-schema.name="Azure CLI" \
+      org.label-schema.version=$CLI_VERSION \
+      org.label-schema.license="MIT" \
+      org.label-schema.description="The Azure CLI is used for all Resource Manager deployments in Azure." \
+      org.label-schema.url="https://docs.microsoft.com/cli/azure/overview" \
+      org.label-schema.usage="https://docs.microsoft.com/cli/azure/install-az-cli2#docker" \
+      org.label-schema.build-date=$BUILD_DATE \
+      org.label-schema.vcs-url="https://github.com/Azure/azure-cli.git" \
+      org.label-schema.docker.cmd="docker run -v \${HOME}/.azure:/root/.azure -it mcr.microsoft.com/azure-cli:$CLI_VERSION"
+
+COPY --from=builder /azure-cli/az.completion /root/.bashrc
+COPY --from=builder /usr/local /usr/local
+COPY --from=tools /usr/local/bin/jp /usr/local/bin/
```

Comment:
I think this is ambiguous and hard to maintain.
Alternatively, we can modify the build script to install and delete the src in one step.

Reply:
I think that is strictly worse, but I see the point. We can move jp into the builder step and then we have just two COPY commands that would have very little reason for change. We could also do this:

```
COPY --from=builder /usr/local /usr/local
COPY --from=tools /usr/local /usr/local
```

I do not think az.completion should be included in the environment to begin with, nor should jp, but here we are. I only want to make the existing solution smaller. Without sacrificing readability or maintainability.

Reply:
I do not think that az.competion.sh should be in /root is what I mean. The jp binary is a single static executable. If the user wants it inside az-cli they can just inject it at container creation or have it mounted. There is no need to have it bundled with az unless there is a runtime dependency to it from az cli itself.

![image](https://user-images.githubusercontent.com/6200749/217097903-370f29f0-c902-4367-9414-d1bdb1d8b9c1.png)

Reply:
I think this line is not safe `COPY --from=builder /usr/local /usr/local`. What if the installed package also modifies another folder. We can't guarantee that everything we need in that folder. (Although it might be true.) 
I prefer using a single RUN command to do the same thing. Like the [CPython dockerfile](https://github.com/docker-library/python/blob/20c8782b7f72091658c18a7347956ec7f7c4010e/3.11/alpine3.17/Dockerfile#L26)

Reply:
In this multistage solution, we do not carry over *any* potential residual components from the build process. That is very hard to achieve with RUN in a single stage. There is always junk left behind and not to mention it is hard to read unmaintainable. There is also observational bias to this commonality of long-winded single-stage builds: nobody dares to change anything because they do not understand the consequences of their changes. Not the best practice, that is not a place where we want to be.

At first, that can seem like a reasonable fear that a generic install script could install something outside the defined install location. To overcome our fear it is enough to open up the install script to see what it does: pip install exclusively. While it could be possible for a setup.py to do something incredibly stupid here the python pip framework is reasonably opinionated on what can and should be done (too opinionated if you ask me). If any files are installed they will be put in the target specified by pip configuration.

Do you feel strong about the /usr/local copy I might be able to fix that. But if you are against the multi-stage I cannot fix that. That is a requirement for a better more maintainable Dockerfile with a smaller build artifact that also can be generated more quickly.

Reply:
I'm not against mult-stage build. If CLI is a Node or Go project, I definitely choose the multi-stage build. I'm sure everything I need is the `dist` folder or the binary.

The new dockerfile also remove `bash`, `git` and other packages. If we choose multi-stage build, we also need to figure out what file is changed during installation, which is more difficult than pip installation.

PS: I know we should not keep irrelevant packages into image. In order to prevent issues like #24836, we still keep them for now.

Reply:
Let me know if you would prefer this kind of solution instead.

https://github.com/hholst80/azure-cli/tree/dockerfile-multistage-alt